File: 20230422_Saturday/20230422.rb
# Leetcode: 1312. Minimum Insertion Steps to Make a String Palindrome.
# https://leetcode.com/problems/minimum-insertion-steps-to-make-a-string-palindrome/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 1474 ms, faster than 100.00% of Ruby online submissions for Minimum
# Insertion Steps to Make a String Palindrome.
# Memory Usage: 224.7 MB, less than 100.00% of Ruby online submissions for
# Minimum Insertion Steps to Make a String Palindrome.
# 2023.04.22 Daily Challenge.
# @param {String} s
# @return {Integer}
def min_insertions(s)
  # Based on the Leetcode's solution.
  n = s.size
  sr = s.reverse
  return n - lcs(s,sr,n)
end
def lcs(s1,s2,n)
  dp = Array.new(n+1){ Array.new(n+1,0)}
  (0..n).each do |i|
    (0..n).each do |j|
      if 0 == i || j == 0
          dp[i][j] = 0
      elsif s1[i-1] == s2[j-1]
        dp[i][j] = 1 + dp[i-1][j-1]
      else
        dp[i][j] = [dp[i-1][j],dp[i][j-1]].max
      end
    end
  end
  return dp[n][n]
end

File: 20230422_Saturday/20230422.go
// Leetcode: 1312. Minimum Insertion Steps to Make a String Palindrome.
// https://leetcode.com/problems/minimum-insertion-steps-to-make-a-string-palindrome/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 16 ms, faster than 66.67% of Go online submissions for Minimum
// Insertion Steps to Make a String Palindrome.
// Memory Usage: 9.5 MB, less than 33.33% of Go online submissions for Minimum
// Insertion Steps to Make a String Palindrome.
// 2023.04.22 Daily Challenge.
package main

import "fmt"

func minInsertions(s string) int {
	// Based on the Leetcode's solution.
	n := len(s)
	sr := ""
	for _, x := range s {
		sr = string(x) + sr
	}
	return n - lcs(s, sr, n)
}

func lcs(s1, s2 string, n int) int {
	dp := make([][]int, n+1)
	for i := 0; i < n+1; i += 1 {
		dp[i] = make([]int, n+1)
	}
	for i := 0; i <= n; i += 1 {
		for j := 0; j <= n; j += 1 {
			if 0 == i || j == 0 {
				dp[i][j] = 0
			} else if s1[i-1] == s2[j-1] {
				dp[i][j] = 1 + dp[i-1][j-1]
			} else {
				dp[i][j] = max(dp[i-1][j], dp[i][j-1])
			}
		}
	}
	return dp[n][n]
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
